src/main/java/com/openlattice/datastore/permissions/controllers/PermissionsController.java

```diff
@@ -125,50 +117,117 @@ public Acl getAcl( @RequestBody AclKey aclKeys ) {
 
-                    for ( Principal user : users ) {
+                    for ( Principal user : users ) { //can add a new ace to a principal that is already in the multimap
                         resultMap.put( user, ace );
                     }
                 }
             }
 
             // compute total permission of each user
             Iterable<AceExplanation> explanation = Iterables.transform( resultMap.asMap().entrySet(),
-                    this::computeAceExplanation );
+                    this::computeAceExplanation ); // passes entry in as key to collection of all aces
             return new AclExplanation( aclKey, explanation::iterator );
         } else {
             throw new ForbiddenException( "Only owner of a securable object can access other users' access rights." );
         }
     }
 
-    /**
-     * Compute the total permission of a user has from his aces, thus computing the Ace explanation
-     */
+
+    // Compute the total permission of a user has from his aces, thus computing the Ace explanation
+
+    private AceExplanation computeAceExplanation( Entry<Principal, Collection<Ace>> entry ) {
+        Set<Permission> totalPermissions = entry.getValue().stream().flatMap( ace -> ace.getPermissions().stream() )
+                .collect( Collectors.toCollection( () -> EnumSet.noneOf( Permission.class ) ) );
+        //OffsetDateTime expirationDate = entry.getValue().stream().flatMap(ace -> ace.getExpirationDate()).collect(  );
+        Set<Ace> aces = entry.getValue().stream().collect( Collectors.toSet() );
+        Ace totalAce = new Ace( entry.getKey(), totalPermissions );
+        return new AceExplanation( totalAce, aces ); //list of securablePrincipals with expiration date
+    }
+    */
+
+    @Override
+    @RequestMapping(
+            path = { EXPLAIN },
+            method = RequestMethod.POST,
+            consumes = MediaType.APPLICATION_JSON_VALUE,
+            produces = MediaType.APPLICATION_JSON_VALUE )
+    public Map<Principal, List<List<Principal>>> getAclExplanation( @RequestBody AclKey aclKey ) {
+        //needs to return a map of principles to the path that grants permission with expiration
+        ensureOwnerAccess( aclKey );
+        SetMultimap<Principal, Ace> resultMap = HashMultimap
+                .create(); //will hold all aces that contain the key principal
+
+        Iterable<Ace> aces = authorizations.getAllSecurableObjectPermissions( aclKey )
+                .getAces(); //gets aces from returned acl
+        Map<PrincipalType, List<Ace>> acesByType = StreamUtil.stream( aces )
+                .collect( Collectors.groupingBy( ace -> ace.getPrincipal().getType() ) );
+        //for principals that aren't users, sPM.getallPrincipalsbyPrincipal gets all of the parents of those principals
+        Map<Principal, List<List<Principal>>> principalToPrincipalPaths = new HashMap<>();
+        //for all principals with direct access
+        for ( Ace ace : aces ) {
```

Comment:
for( Entry<Principal, List<List<Principal>> e : principalToPrincipalPaths ) {
 var parentLayer = securePrincipalsManager.getParentPrincipalsOfPrincipal( ... );
 for( SecurablePrincipal parent : parentLayer ) {
    principalsToPrincipalPaths.merge( parent.getPrincipal(), e.getValue().stream().map( pl -> pl.add(e.getKey() ) ).collect(Collectors.toList()), (p1,p2)->{p1.addAll(p2); return p1;} );

you might also have to move switch Set<List<Principal>> since duplicate paths might get added. It's not great, but haven't had time to look at it closer.

Reply:
Won't the result of calling parent.getPrincipal() return keys that aren't in the map yet?